src/FileSystemActions.ts

```diff
@@ -0,0 +1,154 @@
+                }
+
+                FileSystemActions._eventRegistered = true;
+            }
+        }
+
+        this._disposable = Disposable.from.apply(this, subscriptions);
+    }
+
+    public dispose() {
+        this._disposable.dispose();
+    }
+
+    private static onWillSaveFile(
+        doc: TextDocument,
+        reason: TextDocumentSaveReason
+    ): Promise<boolean> {
+        if (
+            FileSystemActions._lastSavedFileUri?.fsPath === doc.uri.fsPath &&
+            reason !== TextDocumentSaveReason.Manual
+        ) {
+            // don't keep trying when auto-saving (e.g. if the file isn't intended for perforce)
+            return Promise.resolve(true);
+        } else {
+            FileSystemActions._lastSavedFileUri = doc.uri;
+            return FileSystemActions.tryEditFile(doc.uri);
+        }
+    }
+
+    private static onFileModified(docChange: TextDocumentChangeEvent) {
+        const docUri = docChange.document.uri;
+
+        if (docUri.scheme !== "file") {
+            return;
+        }
+
+        //If this doc has already been checked, just returned
+        if (
+            FileSystemActions._lastCheckedFileUri &&
+            docUri.toString() === FileSystemActions._lastCheckedFileUri.toString()
+        ) {
+            return;
+        }
+
+        //Only try to open files open in the editor
+        const editor = window.activeTextEditor;
+        if (
+            !editor ||
+            !editor.document ||
+            editor.document.uri.toString() !== docUri.toString()
+        ) {
+            return;
+        }
+
+        FileSystemActions._lastCheckedFileUri = docUri;
+        FileSystemActions.tryEditFile(docUri);
+    }
+
+    // Had to streamline this, since `onWillSaveTextDocument` allows to delay
+    // the save operation only for a few seconds. The time limit should be
+    // configurable and/or the operation should be cancellable
+    // in future releases.
+    // https://github.com/stef-levesque/vscode-perforce/issues/110
+    private static tryEditFile(uri: Uri): Promise<boolean> {
+        if (
+            PerforceSCMProvider.hasOpenFile(uri) &&
+            !PerforceSCMProvider.mayHaveConflictForFile(uri)
+        ) {
+            return Promise.resolve(true);
+        } else {
+            return PerforceCommands.p4edit(uri);
+        }
+    }
+
+    private static onFilesDeleted(filesDeleted: FileDeleteEvent) {
+        for (const uri of filesDeleted.files) {
+            const fileExcludes = Object.keys(workspace.getConfiguration("files").exclude);
+
+            const shouldIgnore: boolean = micromatch.isMatch(uri.fsPath, fileExcludes, {
+                dot: true
+            });
+            if (!shouldIgnore) {
+                // revert before delete in case it's optn for add/edit.  At
+                // come point maybe dialog to warn user but this does
+                // match logic in PerforceCommands
+                PerforceCommands.p4revert(uri);
```

Comment:
we still need to `await` the `p4revert` call otherwise it will kick off both commands in the same stack frame (since async functions always return promises)

For consistency we should also await the `p4delete` so that onFilesDeleted is now an async function that only resolves when it's finished its job (not that it's being used as such, just as good practice)

Reply:
ok - yes, this makes sense.  I need to think through though if that will work well with the delete events received from the editor. Maybe it's fine.  But I do recall reading something online (maybe with the add or save functions - I don't remember) that they needed to complete within a specific amount of time.  I'll look into it further though.

Reply:
I think that is for `onWillSaveTextDocument` - and only when you explicitly choose to delay the file save - because they need to ensure that saves happen within a reasonable time period As `onDidDeleteFile` is after a delete has already happened I don't think it matters.

Reply:
I have these fixes done, but am having trouble testing.  Have you ever had an issue where running Launch Extension just sits and doesn't actually load a new instance of vscode?  My taskbar is just sitting at the 'building' stage with the throbber spinning and the new instance never spins up.